File: Example/PrebidDemoRenderingKotlin/src/androidTest/java/com/openx/internal_test_app/uiAutomator/utils/RetryRule.java
package com.openx.internal_test_app.uiAutomator.utils;

import org.junit.rules.TestRule;
import org.junit.runner.Description;
import org.junit.runners.model.Statement;

public class RetryRule implements TestRule {
    private int mMaxTries;

    public RetryRule(int maxTries) {
        mMaxTries = maxTries;
    }

    public Statement apply(Statement base, Description description) {
        return statement(base, description);
    }

    private Statement statement(final Statement base, final Description description) {
        return new Statement() {
            @Override
            public void evaluate() throws Throwable {
                Throwable caughtThrowable = null;

                // implement retry logic here
                for (int i = 0; i < mMaxTries; i++) {
                    try {
                        base.evaluate();
                        return;
                    }
                    catch (Throwable t) {
                        caughtThrowable = t;
                        System.err.println(description.getDisplayName() + ": run " + (i + 1) + " failed");
                    }
                }
                System.err.println(description.getDisplayName() + ": giving up after " + mMaxTries + " failures");
                throw caughtThrowable;
            }
        };
    }
}

File: Example/PrebidDemoKotlin/src/androidTest/java/org/prebid/mobile/prebidkotlindemo/utils/RetryRule.kt
package org.prebid.mobile.prebidkotlindemo.utils

import org.junit.rules.TestRule
import org.junit.runner.Description
import org.junit.runners.model.Statement

class RetryRule(private val retryCount: Int): TestRule {

    override fun apply(base: Statement?, description: Description?): Statement = statement(base,description)

    private fun statement(base: Statement?, description: Description?): Statement {
        return object : Statement() {
            override fun evaluate() {
                var caughtThrowable: Throwable? = null
                for (i in 0 until retryCount) {
                    try {
                        base?.evaluate()
                        return
                    } catch (t: Throwable) {
                        caughtThrowable = t
                        System.err.println(description?.displayName + ": run " + (i + 1) + " failed.")
                    }
                }
                System.err.println(description?.displayName + ": Giving up after " + retryCount + " failures.")
                throw caughtThrowable!!
            }
        }
    }
}